src/project.js

```diff
@@ -0,0 +1,705 @@
+  // Prefer the following, which evaluates to a {Promise} that resolves to an
+  // {Array} of {Repository} objects:
+  // ```
+  // Promise.all(atom.project.getDirectories().map(
+  //     atom.project.repositoryForDirectory.bind(atom.project)))
+  // ```
+  getRepositories () {
+    return this.repositories
+  }
+
+  // Public: Get the repository for a given directory asynchronously.
+  //
+  // * `directory` {Directory} for which to get a {Repository}.
+  //
+  // Returns a {Promise} that resolves with either:
+  // * {Repository} if a repository can be created for the given directory
+  // * `null` if no repository can be created for the given directory.
+  repositoryForDirectory (directory) {
+    const pathForDirectory = directory.getRealPathSync()
+    let promise = this.repositoryPromisesByPath.get(pathForDirectory)
+    if (!promise) {
+      const promises = this.repositoryProviders.map((provider) =>
+        provider.repositoryForDirectory(directory)
+      )
+      promise = Promise.all(promises).then((repositories) => {
+        const repo = repositories.find((repo) => repo != null) || null
+
+        // If no repository is found, remove the entry for the directory in
+        // @repositoryPromisesByPath in case some other RepositoryProvider is
+        // registered in the future that could supply a Repository for the
+        // directory.
+        if (repo == null) this.repositoryPromisesByPath.delete(pathForDirectory)
+
+        if (repo && repo.onDidDestroy) {
+          repo.onDidDestroy(() => this.repositoryPromisesByPath.delete(pathForDirectory))
+        }
+
+        return repo
+      })
+      this.repositoryPromisesByPath.set(pathForDirectory, promise)
+    }
+    return promise
+  }
+
+  /*
+  Section: Managing Paths
+  */
+
+  // Public: Get an {Array} of {String}s containing the paths of the project's
+  // directories.
+  getPaths () {
+    return this.rootDirectories.map((rootDirectory) => rootDirectory.getPath())
+  }
+
+  // Public: Set the paths of the project's directories.
+  //
+  // * `projectPaths` {Array} of {String} paths.
+  // * `options` An optional {Object} that may contain the following keys:
+  //   * `mustExist` If `true`, throw an Error if any `projectPaths` do not exist. Any remaining `projectPaths` that
+  //     do exist will still be added to the project. Default: `false`.
+  //   * `exact` If `true`, only add a `projectPath` if it names an existing directory. If `false` and any `projectPath`
+  //     is a file or does not exist, its parent directory will be added instead. Default: `false`.
+  setPaths (projectPaths, options = {}) {
+    for (let repository of this.repositories) {
+      if (repository != null) repository.destroy()
+    }
+    this.rootDirectories = []
+    this.repositories = []
+
+    for (let path in this.watcherPromisesByPath) {
+      this.watcherPromisesByPath[path].then(watcher => { watcher.dispose() })
+    }
```

Comment:
The CoffeeScript version of this code looked like this:

https://github.com/atom/atom/blob/8952cd315d44b7730ed686c2be14a9dc32d1e529/src/project.coffee#L235

[`decaffeinate`](https://www.npmjs.com/package/decaffeinate) translated that to the following JavaScript:

https://github.com/atom/atom/blob/25b7ddb328ab396e7ed53cad0775fdd0eba5ac74/src/project.js#L294

That seems like a non-equivalent translation. `watcherPromisesByPath` is [initialized as an object literal](https://github.com/atom/atom/blob/8952cd315d44b7730ed686c2be14a9dc32d1e529/src/project.coffee#L236), so `watcherPromisesByPath.length` is `undefined`. 

To address this issue, 5937b95b4969c89689fbbfdec3f77e40a01b9eb9 replaces `decaffeinate`'s translation with the `for` loop shown above.

Reply:
:+1: Your translation is spot-on to the intention of the original. It looks like I used an [`in` comprehension when I should have used an `of` loop](http://coffeescript.org/#loops) in the original!